File: src/engine/solarEngine.ts
// src/engine/solarEngine.ts
import { getSunrise, getSunset } from "sunrise-sunset-js"

export type SolarWindow = {
  sunrise: Date | null
  sunset: Date | null
}

export type SolarSplitMode = "solar" | "unverified"

export type DayNightSplit = {
  dayHours: number
  nightHours: number
  mode: SolarSplitMode
}

export type SolarMode = "day" | "night" | "unverified"

export type DriveSegments = {
  type: "Day Only" | "Night Only" | "Mixed Drive" | "Unknown"
  nightStartMs: number | null
  nightEndMs: number | null
  dayStartMs: number | null
  dayEndMs: number | null
}

const THIRTY_MIN_MS = 30 * 60 * 1000

/**
 * Guard against multi-day drives. splitDriveBySolar and computeDayNightSplit
 * can only represent ONE contiguous night range and ONE contiguous day
 * range per drive. A drive spanning more than 24 hours can cross multiple
 * day/night boundaries, which would silently collapse distinct night
 * periods into one incorrect range. Rather than support multi-range data
 * (unnecessary complexity for real teen driving patterns), sessions that
 * exceed this duration are treated as unverified/invalid so the app never
 * displays corrupted solar data. See activeDriveStore.ts for the runtime
 * auto-pause guard that prevents sessions from ever reaching this length.
 */
export const MAX_DRIVE_DURATION_MS = 24 * 60 * 60 * 1000

function isValidDate(value: unknown): value is Date {
  return value instanceof Date && Number.isFinite(value.getTime())
}

function isValidCoordinates(latitude: number, longitude: number): boolean {
  return (
    Number.isFinite(latitude) &&
    Number.isFinite(longitude) &&
    latitude >= -90 &&
    latitude <= 90 &&
    longitude >= -180 &&
    longitude <= 180
  )
}

function getOverlapMs(
  rangeStart: number,
  rangeEnd: number,
  windowStart: number,
  windowEnd: number
): number {
  return Math.max(
    0,
    Math.min(rangeEnd, windowEnd) - Math.max(rangeStart, windowStart)
  )
}

/**
 * Fix for sunrise-sunset-js UTC date-boundary bug:
 * Always anchor the lookup date to LOCAL NOON so the library
 * cannot accidentally resolve sunrise/sunset for the wrong day.
 */
function toSolarLookupDate(date: Date): Date {
  const local = new Date(date)
  local.setHours(12, 0, 0, 0)
  return local
}

/**
 * Civil night ends 30 minutes before sunrise (dawn boundary).
 */
function getCivilNightEndMs(sunriseMs: number): number {
  return sunriseMs - THIRTY_MIN_MS
}

/**
 * Civil night starts 30 minutes after sunset (dusk boundary).
 */
function getCivilNightStartMs(sunsetMs: number): number {
  return sunsetMs + THIRTY_MIN_MS
}

/**
 * Night = before civil dawn OR after civil dusk.
 */
function isBeforeDawnOrAfterDusk(
  timestampMs: number,
  civilNightEndMs: number,
  civilNightStartMs: number
): boolean {
  return timestampMs < civilNightEndMs || timestampMs >= civilNightStartMs
}

/**
 * Returns the sunrise and sunset for the supplied coordinates and local date.
 */
export function getSolarWindowForDate(
  latitude: number,
  longitude: number,
  date: Date = new Date()
): SolarWindow {
  if (!isValidCoordinates(latitude, longitude) || !isValidDate(date)) {
    return { sunrise: null, sunset: null }
  }

  try {
    const lookupDate = toSolarLookupDate(date)

    const rawSunrise = getSunrise(latitude, longitude, lookupDate)
    const rawSunset = getSunset(latitude, longitude, lookupDate)

    const sunrise = isValidDate(rawSunrise) ? rawSunrise : null
    const sunset = isValidDate(rawSunset) ? rawSunset : null

    if (!sunrise || !sunset || sunset.getTime() <= sunrise.getTime()) {
      return { sunrise: null, sunset: null }
    }

    return { sunrise, sunset }
  } catch {
    return { sunrise: null, sunset: null }
  }
}

/**
 * Determines whether a specific timestamp falls during civil darkness.
 */
export function isNightDrive(
  timestamp: Date,
  solarWindow: SolarWindow
): boolean {
  if (!isValidDate(timestamp)) return false

  const { sunrise, sunset } = solarWindow
  if (!isValidDate(sunrise) || !isValidDate(sunset)) return false
  if (sunset.getTime() <= sunrise.getTime()) return false

  const civilNightEndMs = getCivilNightEndMs(sunrise.getTime())
  const civilNightStartMs = getCivilNightStartMs(sunset.getTime())

  return isBeforeDawnOrAfterDusk(
    timestamp.getTime(),
    civilNightEndMs,
    civilNightStartMs
  )
}

/**
 * Gets the current day/night state from solar times only.
 */
export function getCurrentSolarMode(
  date: Date,
  latitude: number,
  longitude: number
): SolarMode {
  if (!isValidDate(date)) return "unverified"

  const solarWindow = getSolarWindowForDate(latitude, longitude, date)
  const { sunrise, sunset } = solarWindow

  if (!sunrise || !sunset) return "unverified"

  return isNightDrive(date, solarWindow) ? "night" : "day"
}

/**
 * Splits one same-calendar-day drive into daylight and darkness hours.
 */
export function computeDayNightSplit(
  startTime: Date,
  endTime: Date,
  solarWindowForDate: (d: Date) => SolarWindow
): DayNightSplit {
  if (!isValidDate(startTime) || !isValidDate(endTime)) {
    return { dayHours: 0, nightHours: 0, mode: "unverified" }
  }

  const startMs = startTime.getTime()
  const endMs = endTime.getTime()
  if (endMs <= startMs) {
    return { dayHours: 0, nightHours: 0, mode: "unverified" }
  }

  // Guard: a drive exceeding 24 hours cannot be safely represented by this
  // single-pass accumulator without risking misclassification across
  // multiple day/night cycles. Treat as unverified rather than guess.
  if (endMs - startMs > MAX_DRIVE_DURATION_MS) {
    return { dayHours: 0, nightHours: 0, mode: "unverified" }
  }

  let cursor = new Date(startTime)
  let dayMs = 0
  let nightMs = 0
  let anyUnverified = false

  while (cursor.getTime() < endMs) {
    const segmentDate = new Date(cursor)
    segmentDate.setHours(12, 0, 0, 0) // anchor to local noon

    const solarWindow = solarWindowForDate(segmentDate)
    const { sunrise, sunset } = solarWindow

    const nextMidnight = new Date(segmentDate)
    nextMidnight.setHours(24, 0, 0, 0)
    const segmentEnd = Math.min(nextMidnight.getTime(), endMs)

    const segStart = cursor.getTime()
    const segEnd = segmentEnd

    if (!isValidDate(sunrise) || !isValidDate(sunset) || sunset.getTime() <= sunrise.getTime()) {
      // No reliable solar data: treat entire segment as night, but mark unverified
      nightMs += segEnd - segStart
      anyUnverified = true
      cursor = new Date(segmentEnd)
      continue
    }

    const sunriseMs = sunrise.getTime()
    const sunsetMs = sunset.getTime()
    const civilNightEndMs = getCivilNightEndMs(sunriseMs)
    const civilNightStartMs = getCivilNightStartMs(sunsetMs)

    const dayOverlap = getOverlapMs(segStart, segEnd, civilNightEndMs, civilNightStartMs)
    const nightOverlap = (segEnd - segStart) - dayOverlap

    dayMs += dayOverlap
    nightMs += nightOverlap

    cursor = new Date(segmentEnd)
  }

  return {
    dayHours: dayMs / 3_600_000,
    nightHours: nightMs / 3_600_000,
    mode: anyUnverified ? "unverified" : "solar",
  }
}

/**
 * Segment-level split for UI: returns actual timestamps for night/day portions.
 * Walks day-by-day so multi-day drives get correct sunrise/sunset per date,
 * instead of relying on one static solar window for the whole drive.
 *
 * NOTE: this function's data model only supports ONE contiguous night range
 * and ONE contiguous day range. It is only correct for drives that cross at
 * most one day/night boundary. The MAX_DRIVE_DURATION_MS guard below
 * prevents it from ever being called with a drive long enough to cross
 * multiple boundaries, which would otherwise silently collapse distinct
 * night periods into a single incorrect range.
 */
export function splitDriveBySolar(
  startTime: Date,
  endTime: Date,
  solarWindowForDate: (d: Date) => SolarWindow
): DriveSegments {
  const invalid: DriveSegments = {
    type: "Unknown",
    nightStartMs: null,
    nightEndMs: null,
    dayStartMs: null,
    dayEndMs: null,
  }

  if (!isValidDate(startTime) || !isValidDate(endTime)) return invalid

  const startMs = startTime.getTime()
  const endMs = endTime.getTime()
  if (endMs <= startMs) return invalid

  // Guard: refuse to compute a single night/day range for a drive long
  // enough to cross multiple solar boundaries. See note above.
  if (endMs - startMs > MAX_DRIVE_DURATION_MS) return invalid

  let cursor = new Date(startTime)
  let nightStartMs: number | null = null
  let nightEndMs: number | null = null
  let dayStartMs: number | null = null
  let dayEndMs: number | null = null
  let hasDay = false
  let hasNight = false

  while (cursor.getTime() < endMs) {
    const segmentDate = new Date(cursor)
    segmentDate.setHours(12, 0, 0, 0)

    const solarWindow = solarWindowForDate(segmentDate)
    const { sunrise, sunset } = solarWindow

    const nextMidnight = new Date(segmentDate)
    nextMidnight.setHours(24, 0, 0, 0)
    const segmentEnd = Math.min(nextMidnight.getTime(), endMs)

    const segStart = cursor.getTime()
    const segEnd = segmentEnd

    if (!isValidDate(sunrise) || !isValidDate(sunset) || sunset.getTime() <= sunrise.getTime()) {
      hasNight = true
      nightStartMs = nightStartMs === null ? segStart : nightStartMs
      nightEndMs = segEnd
      cursor = new Date(segmentEnd)
      continue
    }

    const civilNightEndMs = getCivilNightEndMs(sunrise.getTime())
    const civilNightStartMs = getCivilNightStartMs(sunset.getTime())

    if (segStart < civilNightEndMs) {
      const pieceEnd = Math.min(segEnd, civilNightEndMs)
      hasNight = true
      nightStartMs = nightStartMs === null ? segStart : nightStartMs
      nightEndMs = pieceEnd
    }

    const dayPieceStart = Math.max(segStart, civilNightEndMs)
    const dayPieceEnd = Math.min(segEnd, civilNightStartMs)
    if (dayPieceEnd > dayPieceStart) {
      hasDay = true
      dayStartMs = dayStartMs === null ? dayPieceStart : dayStartMs
      dayEndMs = dayPieceEnd
    }

    if (segEnd > civilNightStartMs) {
      const pieceStart = Math.max(segStart, civilNightStartMs)
      hasNight = true
      nightStartMs = nightStartMs === null ? pieceStart : nightStartMs
      nightEndMs = segEnd
    }

    cursor = new Date(segmentEnd)
  }

  if (hasNight && hasDay) {
    return {
      type: "Mixed Drive",
      nightStartMs,
      nightEndMs,
      dayStartMs,
      dayEndMs,
    }
  }

  if (hasNight) {
    return {
      type: "Night Only",
      nightStartMs,
      nightEndMs,
      dayStartMs: null,
      dayEndMs: null,
    }
  }

  if (hasDay) {
    return {
      type: "Day Only",
      nightStartMs: null,
      nightEndMs: null,
      dayStartMs,
      dayEndMs,
    }
  }

  return invalid
}

/**
 * Final safety net — ensures UI label always matches the computed hours.
 */
export function classifyDriveType(
  dayHours: number,
  nightHours: number
): "Day Only" | "Night Only" | "Mixed Drive" {
  const EPSILON = 0.01

  const hasDay = dayHours > EPSILON
  const hasNight = nightHours > EPSILON

  if (hasDay && hasNight) return "Mixed Drive"
  if (hasNight) return "Night Only"
  if (hasDay) return "Day Only"

  return "Day Only" // safe default — near-zero-duration drives shouldn't imply night
}

File: android/app/src/main/java/com/njdrive50/app/db/SolarEngine.kt
package com.njdrive50.app.db

import android.content.Context
import java.util.Calendar
import java.util.Date
import java.util.TimeZone
import kotlin.math.*

data class SolarWindow(
    val sunrise: Date?,
    val sunset: Date?
)

data class DayNightSplit(
    val dayHours: Double,
    val nightHours: Double,
    val mode: String // "solar" or "unverified"
)

object SolarEngine {

    // CONFIRM: your TS engine and NJDrive50 requirements have referenced
    // both 30 and 40 minutes in different places. Both engines MUST match.
    private const val NIGHT_OFFSET_MS = 30 * 60 * 1000L

    private fun isValidDate(value: Date?): Boolean {
        return value != null
    }

    private fun isValidCoordinates(latitude: Double, longitude: Double): Boolean {
        return latitude.isFinite() &&
                longitude.isFinite() &&
                latitude in -90.0..90.0 &&
                longitude in -180.0..180.0
    }

    private fun getOverlapMs(
        rangeStart: Long,
        rangeEnd: Long,
        windowStart: Long,
        windowEnd: Long
    ): Long {
        return max(0L, min(rangeEnd, windowEnd) - max(rangeStart, windowStart))
    }

    private fun toSolarLookupDate(date: Date): Date {
        val cal = Calendar.getInstance(TimeZone.getDefault())
        cal.time = date
        cal.set(Calendar.HOUR_OF_DAY, 12)
        cal.set(Calendar.MINUTE, 0)
        cal.set(Calendar.SECOND, 0)
        cal.set(Calendar.MILLISECOND, 0)
        return cal.time
    }

    private fun getCivilNightEndMs(sunriseMs: Long): Long {
        return sunriseMs - NIGHT_OFFSET_MS
    }

    private fun getCivilNightStartMs(sunsetMs: Long): Long {
        return sunsetMs + NIGHT_OFFSET_MS
    }

    private fun isBeforeDawnOrAfterDusk(
        timestampMs: Long,
        civilNightEndMs: Long,
        civilNightStartMs: Long
    ): Boolean {
        return timestampMs < civilNightEndMs || timestampMs >= civilNightStartMs
    }

    // -----------------------------------------------------------------
    // Sunrise/Sunset calculation
    // Standard NOAA solar position algorithm (same basis as most
    // sunrise/sunset libraries, including sunrise-sunset-js).
    // Returns null/null if the sun does not rise or set that day
    // (polar regions) — not expected for NJ, but handled safely.
    // -----------------------------------------------------------------
    private fun calculateSunriseSunset(
        latitude: Double,
        longitude: Double,
        date: Date
    ): Pair<Date?, Date?> {
        val cal = Calendar.getInstance(TimeZone.getDefault())
        cal.time = date

        val year = cal.get(Calendar.YEAR)
        val month = cal.get(Calendar.MONTH) + 1
        val day = cal.get(Calendar.DAY_OF_MONTH)

        val zenith = 90.833 // official sunrise/sunset zenith (accounts for refraction)

        val sunriseUtcMinutes = calculateSolarEventUtcMinutes(
            year, month, day, latitude, longitude, zenith, isSunrise = true
        )
        val sunsetUtcMinutes = calculateSolarEventUtcMinutes(
            year, month, day, latitude, longitude, zenith, isSunrise = false
        )

        if (sunriseUtcMinutes == null || sunsetUtcMinutes == null) {
            return Pair(null, null)
        }

        val sunrise = utcMinutesToDate(year, month, day, sunriseUtcMinutes)
        val sunset = utcMinutesToDate(year, month, day, sunsetUtcMinutes)

        return Pair(sunrise, sunset)
    }

    private fun calculateSolarEventUtcMinutes(
        year: Int,
        month: Int,
        day: Int,
        latitude: Double,
        longitude: Double,
        zenith: Double,
        isSunrise: Boolean
    ): Double? {
        val n1 = floor(275.0 * month / 9.0)
        val n2 = floor((month + 9.0) / 12.0)
        val n3 = 1.0 + floor((year - 4.0 * floor(year / 4.0) + 2.0) / 3.0)
        val n = n1 - (n2 * n3) + day - 30.0

        val lngHour = longitude / 15.0
        val t = if (isSunrise) n + ((6.0 - lngHour) / 24.0) else n + ((18.0 - lngHour) / 24.0)

        val m = (0.9856 * t) - 3.289

        var l = m + (1.916 * sin(Math.toRadians(m))) +
                (0.020 * sin(Math.toRadians(2 * m))) + 282.634
        l = normalizeDegrees(l)

        var ra = Math.toDegrees(atan(0.91764 * tan(Math.toRadians(l))))
        ra = normalizeDegrees(ra)

        val lQuadrant = floor(l / 90.0) * 90.0
        val raQuadrant = floor(ra / 90.0) * 90.0
        ra = ra + (lQuadrant - raQuadrant)
        ra /= 15.0

        val sinDec = 0.39782 * sin(Math.toRadians(l))
        val cosDec = cos(asin(sinDec))

        val cosH = (cos(Math.toRadians(zenith)) - (sinDec * sin(Math.toRadians(latitude)))) /
                (cosDec * cos(Math.toRadians(latitude)))

        if (cosH > 1.0 || cosH < -1.0) {
            return null // sun never rises or never sets on this date/location
        }

        val h = if (isSunrise) {
            360.0 - Math.toDegrees(acos(cosH))
        } else {
            Math.toDegrees(acos(cosH))
        } / 15.0

        val time = h + ra - (0.06571 * t) - 6.622
        val utcMinutes = ((time - lngHour) * 60.0)

        return normalizeMinutes(utcMinutes)
    }

    private fun normalizeDegrees(value: Double): Double {
        var v = value
        while (v < 0) v += 360.0
        while (v >= 360) v -= 360.0
        return v
    }

    private fun normalizeMinutes(value: Double): Double {
        var v = value
        while (v < 0) v += 1440.0
        while (v >= 1440) v -= 1440.0
        return v
    }

    private fun utcMinutesToDate(year: Int, month: Int, day: Int, utcMinutes: Double): Date {
        val cal = Calendar.getInstance(TimeZone.getTimeZone("UTC"))
        cal.set(year, month - 1, day, 0, 0, 0)
        cal.set(Calendar.MILLISECOND, 0)
        cal.add(Calendar.MINUTE, utcMinutes.toInt())
        cal.add(Calendar.SECOND, ((utcMinutes - utcMinutes.toInt()) * 60).toInt())
        return cal.time
    }

    // -----------------------------------------------------------------
    // Public API — mirrors src/engine/solarEngine.ts
    // -----------------------------------------------------------------

    fun getSolarWindowForDate(
        latitude: Double,
        longitude: Double,
        date: Date = Date()
    ): SolarWindow {
        if (!isValidCoordinates(latitude, longitude)) {
            return SolarWindow(null, null)
        }

        return try {
            val lookupDate = toSolarLookupDate(date)
            val (sunrise, sunset) = calculateSunriseSunset(latitude, longitude, lookupDate)

            if (!isValidDate(sunrise) || !isValidDate(sunset) || sunset!!.time <= sunrise!!.time) {
                SolarWindow(null, null)
            } else {
                SolarWindow(sunrise, sunset)
            }
        } catch (_: Exception) {
            SolarWindow(null, null)
        }
    }

    fun isNightDrive(timestampMs: Long, solarWindow: SolarWindow): Boolean {
        val sunrise = solarWindow.sunrise
        val sunset = solarWindow.sunset

        if (!isValidDate(sunrise) || !isValidDate(sunset)) return false
        if (sunset!!.time <= sunrise!!.time) return false

        val civilNightEndMs = getCivilNightEndMs(sunrise.time)
        val civilNightStartMs = getCivilNightStartMs(sunset.time)

        return isBeforeDawnOrAfterDusk(timestampMs, civilNightEndMs, civilNightStartMs)
    }

    fun isNight(
        context: Context,
        latitude: Double,
        longitude: Double,
        timestampMs: Long
    ): Boolean {
        val date = Date(timestampMs)
        val solarWindow = getSolarWindowForDate(latitude, longitude, date)

        if (!isValidDate(solarWindow.sunrise) || !isValidDate(solarWindow.sunset)) return false

        return isNightDrive(timestampMs, solarWindow)
    }

    fun computeDayNightSplit(
        startTime: Date,
        endTime: Date,
        solarWindowForDate: (Date) -> SolarWindow
    ): DayNightSplit {
        val startMs = startTime.time
        val endMs = endTime.time
        if (endMs <= startMs) {
            return DayNightSplit(0.0, 0.0, "unverified")
        }

        var cursor = Date(startMs)
        var dayMs = 0L
        var nightMs = 0L
        var anyUnverified = false

        while (cursor.time < endMs) {
            val segmentDate = toSolarLookupDate(cursor)
            val solarWindow = solarWindowForDate(segmentDate)
            val sunrise = solarWindow.sunrise
            val sunset = solarWindow.sunset

            val cal = Calendar.getInstance(TimeZone.getDefault())
            cal.time = segmentDate
            cal.set(Calendar.HOUR_OF_DAY, 24)
            cal.set(Calendar.MINUTE, 0)
            cal.set(Calendar.SECOND, 0)
            cal.set(Calendar.MILLISECOND, 0)
            val nextMidnightMs = cal.time.time

            val segmentEnd = min(nextMidnightMs, endMs)
            val segStart = cursor.time
            val segEnd = segmentEnd

            if (!isValidDate(sunrise) || !isValidDate(sunset) || sunset!!.time <= sunrise!!.time) {
                nightMs += segEnd - segStart
                anyUnverified = true
                cursor = Date(segmentEnd)
                continue
            }

            val civilNightEndMs = getCivilNightEndMs(sunrise.time)
            val civilNightStartMs = getCivilNightStartMs(sunset.time)

            val dayOverlap = getOverlapMs(segStart, segEnd, civilNightEndMs, civilNightStartMs)
            val nightOverlap = (segEnd - segStart) - dayOverlap

            dayMs += dayOverlap
            nightMs += nightOverlap

            cursor = Date(segmentEnd)
        }

        val dayHours = dayMs / 3_600_000.0
        val nightHours = nightMs / 3_600_000.0
        val mode = if (anyUnverified) "unverified" else "solar"

        return DayNightSplit(dayHours, nightHours, mode)
    }
}